core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1413,106 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
+        )
+      } yield {
+        def enqueue(t: O): F2[Boolean] =
+          for {
+            _ <- demand.acquire
+            buf <- buffer.modify(buf => (buf.copy(buf.data :+ t), buf))
+            _ <- supply.release
+          } yield buf.endOfDemand.isEmpty
+
+        def waitN(s: Semaphore[F2]) =
+          F.guaranteeCase(s.acquireN(n.toLong)) {
+            case Succeeded(_) => s.releaseN(n.toLong)
+            case _            => F.unit
+          }
+
+        def acquireSupplyUpToNWithin(n: Long, timeout: FiniteDuration): F2[Long] =
+          // in JS cancellation doesn't always seem to run, so race conditions should restore state on their own
+          F.race(
+            F.sleep(timeout),
+            waitN(supply)
+          ).flatMap {
+            case Left(_) =>
+              for {
+                _ <- supply.acquire
+                m <- supply.available
+                k = m.min(n - 1)
+                b <- supply.tryAcquireN(k)
+              } yield if (b) k + 1 else 1
+            case Right(_) =>
+              supply.acquireN(n) *> F.pure(n)
           }
 
-        // Invariants:
-        // acc.size < n, always
-        // hasTimedOut == true iff a timeout has been received, and acc.isEmpty
-        def go(acc: Chunk[O], timedPull: Pull.Timed[F2, O], hasTimedOut: Boolean = false)
-            : Pull[F2, Chunk[O], Unit] =
-          timedPull.uncons.flatMap {
-            case None =>
-              Pull.output1(acc).whenA(acc.nonEmpty)
-            case Some((e, next)) =>
-              def resetTimerAndGo(q: Chunk[O]) =
-                timedPull.timeout(timeout) >> go(q, next)
-
-              e match {
-                case Left(_) =>
-                  if (acc.nonEmpty)
-                    Pull.output1(acc) >> resetTimerAndGo(Chunk.empty)
-                  else
-                    go(Chunk.empty, next, hasTimedOut = true)
-                case Right(c) if hasTimedOut =>
-                  // it has timed out without reset, so acc is empty
-                  val (toEmit, rest) =
-                    if (c.size < n) Pull.output1(c) -> Chunk.empty
-                    else resize(c, Pull.done)
-                  toEmit >> resetTimerAndGo(rest)
-                case Right(c) =>
-                  val newAcc = acc ++ c
-                  if (newAcc.size < n)
-                    go(newAcc, next)
-                  else {
-                    val (toEmit, rest) = resize(newAcc, Pull.done)
-                    toEmit >> resetTimerAndGo(rest)
+        def dequeueN(n: Int): F2[Option[Vector[O]]] =
+          acquireSupplyUpToNWithin(n.toLong, timeout).flatMap { n =>
+            buffer
+              .modify { buf =>
+                if (buf.data.size >= n) {
+                  val (head, tail) = buf.data.splitAt(n.toInt)
+                  (buf.copy(tail), buf.copy(head))
+                } else {
+                  (buf.copy(Vector.empty), buf)
+                }
+              }
```

Comment:
Could we extract this as a method of `JunctionBuffer`?

```Scala
def splitAt(n: Long): (JunctionBuffer[O], JunctionBuffer[O]) = 
  if (n >= data.size) {
    val (head, tail) = data.splitAt(n.toInt)
    (copy(tail), copy(head))
  } else 
    (copy(Vector.empty), this)
``` 

So that this becomes  `buffer.modify(_.splitAt(n))`

Reply:
Done